# Data Science 7 days of code 
## Day 1
### Instruçoes
Neste primeiro dia, eu te proponho baixar um dataset do portal do CEAPS (Cota para Exercício da Atividade Parlamentar dos Senadores) e aplicar processos de tratamento e limpeza de dados nele (processo conhecido como Data Wrangling). Lidar com dados nulos (deletar ou imputar um valor novo, por exemplo); remover colunas que não trazem nenhuma informação; processar datas que estão em formato incorreto; alterar o tipo da coluna (uma coluna que é numérica está como texto no Pandas); remover duplicados; dentre outras.<br>**Instruções do primeiro dia tiradas do e-mail enviados a mim**<br>

## Visalisando os dados
Antes de começar vamos baixar os dados e visuliza-los com a biblioteca *Pandas* do python, para ver como os dados no arquivo *XLSX* estão organizados.

In [697]:
import pandas as pd
import numpy as np

dados = pd.read_excel("despesa_ceaps_2022.xlsx")
dados

ANO  MES  ... VALOR_REEMBOLSADO  COD_DOCUMENTO
0      2022    1  ...              6000        2173614
1      2022    1  ...              1500        2173615
2      2022    1  ...              6000        2173616
3      2022    1  ...              1000        2173618
4      2022    1  ...              2000        2173617
...     ...  ...  ...               ...            ...
16777  2022   12  ...           2893,04        2191398
16778  2022   12  ...           1180,19        2192272
16779  2022   12  ...            2671,9        2192274
16780  2022   12  ...           1334,31        2192244
16781  2022   12  ...           2250,72        2193622

[16782 rows x 11 columns]

## Análise descritiva dos dados

In [698]:
lc = dados.shape
print(f"Linhas --> {lc[0]}\nColunas --> {lc[1]}")

Linhas --> 16782
Colunas --> 11


In [699]:
colunas = dados.columns
print("__________ Nome das colunas __________\n")
for i, e in enumerate(colunas) :
    print(f"{i}. {e}")

__________ Nome das colunas __________

0. ANO
1. MES
2. SENADOR
3. TIPO_DESPESA
4. CNPJ_CPF
5. FORNECEDOR
6. DOCUMENTO
7. DATA
8. DETALHAMENTO
9. VALOR_REEMBOLSADO
10. COD_DOCUMENTO


**Agora irei visualizar o índice do *Data Frame* geral**.

In [700]:
dados.index

RangeIndex(start=0, stop=16782, step=1)

## Lidar com dados nulos

Para remover os dados nulos vou utilizar o método *info* do *Pandas* e identificar as colunas que não tem o mesmo número de valores não nulos.

In [701]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16782 entries, 0 to 16781
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   ANO                16782 non-null  int64 
 1   MES                16782 non-null  int64 
 2   SENADOR            16782 non-null  object
 3   TIPO_DESPESA       16782 non-null  bool  
 4   CNPJ_CPF           16782 non-null  object
 5   FORNECEDOR         16782 non-null  object
 6   DOCUMENTO          15848 non-null  object
 7   DATA               16782 non-null  object
 8   DETALHAMENTO       9062 non-null   object
 9   VALOR_REEMBOLSADO  16782 non-null  object
 10  COD_DOCUMENTO      16782 non-null  int64 
dtypes: bool(1), int64(3), object(7)
memory usage: 1.3+ MB


**Aqui podemos ver que as colunas *DOCUMENTO* e *DETALHAMENTO* possuem entradas diferentes de 16782, logo possuem valores do tipo *None* no python. Para tratar esses dados nulos irei remover utilizando o método *dropna* do *Pandas* nas respectivas colunas.**

In [702]:
dados = dados.dropna(subset=["DOCUMENTO", "DETALHAMENTO", "VALOR_REEMBOLSADO"])
dados

ANO  MES  ... VALOR_REEMBOLSADO  COD_DOCUMENTO
0      2022    1  ...              6000        2173614
1      2022    1  ...              1500        2173615
2      2022    1  ...              6000        2173616
3      2022    1  ...              1000        2173618
4      2022    1  ...              2000        2173617
...     ...  ...  ...               ...            ...
16777  2022   12  ...           2893,04        2191398
16778  2022   12  ...           1180,19        2192272
16779  2022   12  ...            2671,9        2192274
16780  2022   12  ...           1334,31        2192244
16781  2022   12  ...           2250,72        2193622

[8396 rows x 11 columns]

**Agora temos um *Data Frame* sem dados nulos, para verificar isso vou chamar o método *info* de novo para verificar se o número de linhas é igual ou número de entradas não nulas.**

In [703]:
lc = dados.shape
print(f"Linhas --> {lc[0]}\nColunas --> {lc[1]}\n")
print(dados.info())

Linhas --> 8396
Colunas --> 11

<class 'pandas.core.frame.DataFrame'>
Index: 8396 entries, 0 to 16781
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   ANO                8396 non-null   int64 
 1   MES                8396 non-null   int64 
 2   SENADOR            8396 non-null   object
 3   TIPO_DESPESA       8396 non-null   bool  
 4   CNPJ_CPF           8396 non-null   object
 5   FORNECEDOR         8396 non-null   object
 6   DOCUMENTO          8396 non-null   object
 7   DATA               8396 non-null   object
 8   DETALHAMENTO       8396 non-null   object
 9   VALOR_REEMBOLSADO  8396 non-null   object
 10  COD_DOCUMENTO      8396 non-null   int64 
dtypes: bool(1), int64(3), object(7)
memory usage: 729.7+ KB
None


**Podemos observar que agora todas as colunas carregam o mesmo número de linhas que os dados não nulos.**

## Removendo coluna

**Na célula de código abaixo podemos observar que a coluna *TIPO_DESPESA* não traz nada útil para esta *Data Frame*, visto que é uma coluna que possui apenas um valor: *False*, então utilizarei o método *drop* para exculí-la.**

In [704]:
print(dados["TIPO_DESPESA"].unique())

[False]


In [705]:
dados = dados.drop("TIPO_DESPESA", axis=1)
dados

ANO  MES  ... VALOR_REEMBOLSADO COD_DOCUMENTO
0      2022    1  ...              6000       2173614
1      2022    1  ...              1500       2173615
2      2022    1  ...              6000       2173616
3      2022    1  ...              1000       2173618
4      2022    1  ...              2000       2173617
...     ...  ...  ...               ...           ...
16777  2022   12  ...           2893,04       2191398
16778  2022   12  ...           1180,19       2192272
16779  2022   12  ...            2671,9       2192274
16780  2022   12  ...           1334,31       2192244
16781  2022   12  ...           2250,72       2193622

[8396 rows x 10 columns]

**Agora temos um *Data Frame* sem a coluna *TIPO_DESPESA*, pois não tinha nenhuma utilidade.**

## Analisando e tratando colunas 

**Na célula de código abaixo ireiverificar o tipo de variável de cada coluna com o método *d_types*.**

In [706]:
dados.dtypes

ANO                   int64
MES                   int64
SENADOR              object
CNPJ_CPF             object
FORNECEDOR           object
DOCUMENTO            object
DATA                 object
DETALHAMENTO         object
VALOR_REEMBOLSADO    object
COD_DOCUMENTO         int64
dtype: object

**Podemos ver que a maioria das colunas tem valores do tipo *object*, ou seja, muitas entradas tem tipos de dados misturados na mesma coluna, para resolver isso a melhor opção é mudar o tipo de variável para uma só.**<br><br>**Na célula de código abaixo identifiquei na coluna *VALOR_REEMBOLSADO* todas as entradas que eram do tipo *stirng* e onde tinha  ','  foi trocado por um  '.'  para fazer a troca de tipo de variável, que no caso foi trocada para *float*.**

In [707]:
lista = []
strings = []
floats = []

#Identificando e trocando ',' por '.'
for v in dados["VALOR_REEMBOLSADO"] :
    if type(v) == str :
        v = list(v)
        for l in range(len(v)) :
            if v[l] == "," :
                v[l] = "."
                lista.append(v)
                
#Concatenando as strings das listas
for x in lista :
    s = "".join(x)
    strings.append(s)

#Transformando em float
for n in strings :
    n = float(n)
    floats.append(n)

**Na célula de código abaixo irei identificar na coluna *VALOR_REEMBOLSADO* qual valores são do tipo *string* e trocar pelas novas variáveis do tipo *float*, assim substituindo tudo por valores numéricos**

In [708]:
valores = []
novos_valores = []
valores_float = []
ifl = 0

for valor in dados["VALOR_REEMBOLSADO"] :
    valores.append(valor)
for iv in range(len(valores)) :
    if type(valores[iv]) != str :
        novos_valores.append(valores[iv])
    else:
        valores[iv] = floats[ifl]
        novos_valores.append(valores[iv])
        ifl += 1
for q in novos_valores :
    q = float(q)
    valores_float.append(q)

valores_float = np.array(valores_float)

**Agora irei criar uma nova coluna chamada de *VALOR* com apenas entradas numéricas, assim poderei fazer operações matemáticas com ela, também remover a coluna *VALOR_REEMBOLSADO* que é inútil para fazer análises e utilizar o método *drop_duplicates* para ver se há dados repetidos.**

In [709]:
dados["VALOR"] = valores_float
dados = dados.drop(["VALOR_REEMBOLSADO"], axis=1)
dados = dados.drop_duplicates()
dados

ANO  MES           SENADOR  ... DETALHAMENTO COD_DOCUMENTO    VALOR
0      2022    1      ACIR GURGACZ  ...        False       2173614  6000.00
1      2022    1      ACIR GURGACZ  ...        False       2173615  1500.00
2      2022    1      ACIR GURGACZ  ...        False       2173616  6000.00
3      2022    1      ACIR GURGACZ  ...        False       2173618  1000.00
4      2022    1      ACIR GURGACZ  ...        False       2173617  2000.00
...     ...  ...               ...  ...          ...           ...      ...
16777  2022   12  ZEQUINHA MARINHO  ...        False       2191398  2893.04
16778  2022   12  ZEQUINHA MARINHO  ...        False       2192272  1180.19
16779  2022   12  ZEQUINHA MARINHO  ...        False       2192274  2671.90
16780  2022   12  ZEQUINHA MARINHO  ...        False       2192244  1334.31
16781  2022   12  ZEQUINHA MARINHO  ...        False       2193622  2250.72

[8396 rows x 10 columns]

**Fazendo uma verificação das características dos dados novamente.**

In [716]:
lc = dados.shape
print(f"\nLinhas --> {lc[0]}\nColunas --> {lc[1]}\n")

colunas = dados.columns
print("__________ Nome das colunas __________")
for i, e in enumerate(colunas) :
    print(f"{i}. {e}")
print("")
print(dados.info())


Linhas --> 8396
Colunas --> 10

__________ Nome das colunas __________
0. ANO
1. MES
2. SENADOR
3. CNPJ_CPF
4. FORNECEDOR
5. DOCUMENTO
6. DATA
7. DETALHAMENTO
8. COD_DOCUMENTO
9. VALOR

<class 'pandas.core.frame.DataFrame'>
Index: 8396 entries, 0 to 16781
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   ANO            8396 non-null   int64  
 1   MES            8396 non-null   int64  
 2   SENADOR        8396 non-null   object 
 3   CNPJ_CPF       8396 non-null   object 
 4   FORNECEDOR     8396 non-null   object 
 5   DOCUMENTO      8396 non-null   object 
 6   DATA           8396 non-null   object 
 7   DETALHAMENTO   8396 non-null   object 
 8   COD_DOCUMENTO  8396 non-null   int64  
 9   VALOR          8396 non-null   float64
dtypes: float64(1), int64(3), object(6)
memory usage: 721.5+ KB
None


## Dia 1 finalizado 
Tratamos o *Data Frame* retiramos dados nulos e criamos uma nova coluna numérica para operações futuras. Agora salvarei esse arquivo em *CSV* para utilizá-lo depois.<br>**Segue o código abaixo para salvá-lo.**

In [718]:
dados.to_csv("ceaps_day1.csv", index=False)

In [719]:
pd.read_csv("ceaps_day1.csv")

ANO  MES           SENADOR  ... DETALHAMENTO COD_DOCUMENTO    VALOR
0     2022    1      ACIR GURGACZ  ...        False       2173614  6000.00
1     2022    1      ACIR GURGACZ  ...        False       2173615  1500.00
2     2022    1      ACIR GURGACZ  ...        False       2173616  6000.00
3     2022    1      ACIR GURGACZ  ...        False       2173618  1000.00
4     2022    1      ACIR GURGACZ  ...        False       2173617  2000.00
...    ...  ...               ...  ...          ...           ...      ...
8391  2022   12  ZEQUINHA MARINHO  ...        False       2191398  2893.04
8392  2022   12  ZEQUINHA MARINHO  ...        False       2192272  1180.19
8393  2022   12  ZEQUINHA MARINHO  ...        False       2192274  2671.90
8394  2022   12  ZEQUINHA MARINHO  ...        False       2192244  1334.31
8395  2022   12  ZEQUINHA MARINHO  ...        False       2193622  2250.72

[8396 rows x 10 columns]